src/Factory.php

```diff
@@ -127,31 +119,37 @@ protected function buildClaims()
 
         // add the default claims
         foreach ($this->defaultClaims as $claim) {
-            $this->addClaim($claim, $this->$claim());
+            $this->addClaim($claim, $this->claimFactory->make($claim));
         }
 
         // add custom claims on top, allowing them to overwrite defaults
         $this->addClaims($this->getCustomClaims());
 
+        // add the jti last since it is based on all other claims,
+        // but only if it hasn't already been added.
+        if (! $this->claims->has('jti')) {
+            $this->addClaim('jti', $this->jti());
```

Comment:
Possibility: `jti` doesn't need to be a special case.

If the claim factory handles construction of the entire collection (independent question: is there a reason to leave that in the main Factory?), then we could do:
```php
abstract class Claim {
    public static $creationPriority = 0;
    // ...
}
class JwtId extends Claim {
    public static $creationPriority = -100;
    public function setDefault() {
        $this->value = md5(sprintf('%s.%s', $this->factoryInstance->claims->toJson(), Str::quickRandom()));
    }
    // ...
}
class ClaimFactory {
    public function getCollection(array $rawClaims)
    {
        return $collect($rawClaims)->sortByDesc(function ($value, $name) {
            // sort the collection from highest to lowest priority
            return self::$classMap[$name]::$creationPriority;
        })->transform(function ($value, $name) {
            // replace each entry with the claim instance
            return $this->get($name, $value);
        });
    }
    // ...
}
```

That would allow us to guarantee order of creation, so `jti` could still depend on the other values, but avoid a hard-coded special case.

Regardless, I think it might be good to consider letting the claim factory build the whole collection. Just have regular factory handle the way the user provides input (collecting the claims as an array until `make` is called).